### PR TITLE
Configure nginx/uwsgi in clank

### PR DESCRIPTION
**Pull request by cdosborn**
## Description
This pr addresses several issues:
- atmosphere/troposphere have duplicate logic for setting up uwsgi/nginx
- developers should be able to use clank to support development nginx environments (like running `manage.py runserver`, or `webpack-dev-server` behind nginx).
- flower didn't need to manage https (nginx manages it now)
  - flower now runs as www-data rather than root (before it need root to access tls private key)

Nginx configuration now supports preconfigured setups. See the updated clank readme. A developer can now use clank to switch between various setups with ease by running the atmo api locally to throw in a debug statement, or by preferring the webpack-dev-server, etc.

Several small issues were identified in our current nginx/uwsgi installation.

ssl stapling was turned on but was configured improperly. It is now fixed by passing the `ssl_trusted_certificate` nginx directive. This means that browsers can determine if a cert has been revoked (maybe because the private key was stolen), and can reject the connection.

uwsgi autoreload was configured improperly. It is fixed now, and `UWSGI_AUTORELOAD: true` will enable it in a development environment.

This will run the two roles added: `configure-nginx, configure-uwsgi`.
```
ansible-playbook -e @$ENV_FILE playbooks/deploy_stack.yml  --tags ssl,uwsgi,nginx
```
For `nginx` you'll want to try various values for `nginx_locations` and see how `/etc/nginx/sites-enabled/site.conf` gets templated.
For `uwsgi` you'll want to check that the various `/etc/uwsgi/apps-enabled/*` get generated and that starting uwsgi services is :+1: 

Depends on:
- https://github.com/cyverse/clank/pull/249
  - This upgrade of ansible fixes the default behavior of the template module, that this relies on
- https://github.com/cyverse/atmosphere/pull/573
  - Clean atmo/extras folder, and don't generate nginx/uwsgi sections
- https://github.com/cyverse/troposphere/pull/745
  - Clean troposphere/extras folder, and don't generate nginx/uwsgi sections
- https://gitlab.cyverse.org/atmosphere/atmo-local/merge_requests/50
  - Add sensible dev defaults for uwsgi, otherwise uses default nginx variables
- https://gitlab.cyverse.org/atmosphere/atmo-beta/merge_requests/33
  - Add nginx variables to support jenkins
- https://gitlab.cyverse.org/atmosphere/atmo-dev/merge_requests/23
  - Add nginx variables to support jenkins

## Checklist before merging Pull Requests
- [x] New test(s) included to reproduce the bug/verify the feature
- [x] Documentation created/updated (README.md and dist_files/variables.yml.dist)
  - See update [README.md](https://github.com/cyverse/clank/pull/247/files#diff-04c6e90faac2675aa89e2176d2eec7d8)
  - See new configure-uwsgi role [README.md](https://github.com/cyverse/clank/pull/247/files#diff-d57fcdbd90905ec9321240815732c7a2)
  - See new configure-nginx role [README.md](https://github.com/cyverse/clank/pull/247/files#diff-fc2ba26a31b88c845b1f2ce763945b0f)
- [x] Reviewed and approved by at least one other contributor.
- [x] Updated CHANGELOG.md
- [ ] New variables committed to secrets repos

### CODE REVIEW

File: group_vars/common [hunk 1]
```diff
@@ -7,12 +7,8 @@ virtualenv_dir: "{{VIRTUAL_ENV_DIR | default('/opt/env')}}"
 github_branch: master
 pg_version: 9.6
 server_name: "{{ SERVER_NAME | default(ansible_fqdn) }}"
-nginx_combined_cert_dir: "/etc/ssl/certs"
-nginx_combined_cert_file: "combined_atmo_ssl_cert.pem"
-nginx_combined_cert_path: "{{nginx_combined_cert_dir}}/{{nginx_combined_cert_file}}"
-nginx_key_dir: /etc/ssl/private
-nginx_key_file: atmo_ssl_cert.key
-nginx_key_path: "{{nginx_key_dir}}/{{nginx_key_file}}"
+nginx_server_url: "https://{{ server_name }}"
```
Comment (lenards) [inline group_vars/common]: I think this is causing the failure in Travis CI deploy here:
- https://travis-ci.org/cyverse/clank/builds/325238052#L3247

I am not _totally_ **why** is a recursive loop yet, but the error on Line 3247 of the Travis check has:

```
... recursive loop detected in template string: {{ SERVER_NAME | default(ansible_fqdn) }}"
```

It's in the Task `configure-nginx : Template nginx site: /etc/nginx/sites-available/site.conf`

Comment (lenards) [reply group_vars/common]: Ugh ... Travis's linking to the line doesn't seem to always place you at the site of the error. Sorry. I think it's the progressive loading.

Comment (lenards) [reply group_vars/common]: Wow .. it's more difficult to find that failing play/task then I expected.

Comment (lenards) [reply group_vars/common]: HAHAHA ... I was looking in `master`, but it is in this PR.

File: playbooks/deploy_stack.yml [hunk 2]
```diff
@@ -10,4 +10,18 @@
 - include: ./prepare_host.yml
 - include: ./deploy_atmosphere.yml
 - include: ./deploy_troposphere.yml
+
+- name: Configure nginx
+  hosts: full_deploy
+  roles:
+    - { role: configure-nginx,
+        SERVER_URL: "{{ nginx_server_url }}",
+        SERVER_NAME: "{{ server_name }}",
```
Comment (lenards) [inline playbooks/deploy_stack.yml]: I think Line 18, `SERVER_URL` is this where we get the recursive loop, when replacing the symbol `nginx_server_url`, it expanding to:

```
"https://{{ server_name }}"
```

And, `server_name` expands to:
```
server_name: "{{ SERVER_NAME | default(ansible_fqdn) }}"
```

So I think that is where the Travis CI failure comes in.

**Comment by cdosborn** [comment]
I fixed the travis issue by renaming "SERVER_NAME" to "SITE_SERVER_NAME" in the role.

**Comment by cdosborn** [comment]
This pr requires a change to atmosphere/troposphere, so that the configure script will not template nginx/uwsgi.

**Comment by lenards** [comment]
I think it is a matter of just removing variables_templates.json:

https://github.com/cyverse/troposphere/blob/master/variables_templates.json#L4-L5

https://github.com/cyverse/troposphere/blob/master/variables_templates.json#L12-L21

Then `./configure` will not try to run those.

And we'd need to clean up the template files in `extras/`.

**Comment by cdosborn** [comment]
Okay, also realized that jenkins is not supported as a template

File: README.md [hunk 3]
```diff
@@ -140,6 +140,51 @@ File Dimentions
...
+ATMO_DEV_SERVER_URL: "http://local.atmo.cloud:8000"
+```
+
+Note, that `ATMO_DEV_SERVER_URL` must be defined, and you will be responsible
+
+In order to serve jenkins, `nginx_locations` would include `"jenkins"`.
+```
+LOCATIONS: ["jenkins", ...]
+JENKINS_SERVER_URL: "http://127.0.0.1:8080"
+```
```
Comment (lenards) [inline README.md]: This is not important for the review, but it is worth mentioning that I believe the desire not for Atmosphere to host its' own Jenkins CI, but for us to have common, team-wide CI server. 

So I would think if we have clarity of this, we can remove this language at a later time.

File: playbooks/deploy_atmosphere.yml [hunk 4]
```diff
@@ -68,7 +68,7 @@
       tags: [ 'atmosphere', 'setup-virtualenv']
 
 - name: Include setup atmosphere
-  include: ./setup_atmosphere.yml
+  import_playbook: ./setup_atmosphere.yml
```
[the remark this answers is not in the document]
Comment (cdosborn) [reply playbooks/deploy_atmosphere.yml]: Yeah this pr depends on the ansible 2.4 pr linked above, so those commits are shining through here

**Comment by lenards** [comment]
Connor's branch is [successfully building now](https://travis-ci.org/cdosborn/clank/builds/327019531) ✅

**Comment by lenards** [comment]
🎉

**Pull request merged**